cmd/clusterctl/client/cluster/mover.go

```diff
@@ -102,6 +110,100 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
 	return o.move(objectGraph, proxy)
 }
 
+func (o *objectMover) Save(namespace string, directory string) error {
+	log := logf.Log
+	log.Info("Performing save...")
+
+	objectGraph := newObjectGraph(o.fromProxy, o.fromProviderInventory)
+
+	// Gets all the types defines by the CRDs installed by clusterctl plus the ConfigMap/Secret core types.
+	err := objectGraph.getDiscoveryTypes()
+	if err != nil {
+		return err
+	}
+
+	// Discovery the object graph for the selected types:
+	// - Nodes are defined the Kubernetes objects (Clusters, Machines etc.) identified during the discovery process.
+	// - Edges are derived by the OwnerReferences between nodes.
+	if err := objectGraph.Discovery(namespace); err != nil {
+		return err
+	}
+
+	// Checks if Cluster API has already completed the provisioning of the infrastructure for the objects involved in the save operation.
+	// This is required because if the infrastructure is provisioned, then we can reasonably assume that the objects we are saving are
+	// not currently waiting for long-running reconciliation loops, and so we can safely rely on the pause field on the Cluster object
+	// for blocking any further object reconciliation on the source objects.
+	if err := o.checkProvisioningCompleted(objectGraph); err != nil {
+		return err
+	}
+
+	// Check whether nodes are not included in GVK considered for move
+	objectGraph.checkVirtualNode()
+
+	if err := o.save(objectGraph, directory); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *objectMover) Restore(namespace string, glob string, directory string) error {
+	log := logf.Log
+	log.Info("Performing restore...")
+
+	objectGraph := newObjectGraph(o.fromProxy, o.fromProviderInventory)
+
+	// Rebuild the restore objectgraph from files (based on tag / glob / directory)
+	objectsMatcher := regexp.MustCompile(`^.*` + glob + `.*`)
+
+	files, err := filepath.Glob(directory + "/*")
+	if err != nil {
+		return err
+	}
+
+	var rawYamls [][]byte
+	for _, objectFile := range files {
+		if objectsMatcher.MatchString(objectFile) {
+			byObj, err := ioutil.ReadFile(objectFile)
+			if err != nil {
+				return err
+			}
+
+			rawYamls = append(rawYamls, byObj)
+		}
+	}
+
+	processedYamls := yaml.JoinYaml(rawYamls...)
+
+	objs, err := yaml.ToUnstructured(processedYamls)
+	if err != nil {
+		return err
+	}
+
+	// Gets all the types defined by the CRDs provided by file plus the ConfigMap/Secret core types.
+	err = objectGraph.getDiscoveryFileTypes(objs)
```

Comment:
Might be I'm missing something, but why not use getDiscoveryTypes that reads from CRDs?

Reply:
`getDiscoveryTypes` uses a client to read the CRDs off of a management cluster. If we are restoring a management cluster from files, there would be no actual cluster to read CRDs from. Instead, `getDiscoveryFileTypes` _reads CRDs from objects passed in from files_. The intent was to do much of the same thing as `getDiscoveryTypes`, but instead, do it from a grouping of objects from files.

Reply:
According to the save/backup logic in this PR, the files contains only cluster object, machine deployment object etc. but not the actual Cluster API CRDs nor the Cluster API deployments/reach rules/ web hook configurations required to run the controllers.
As a consequence I assume that the something (e.g clusterctl) should install Cluster API (including CRDs) before doing restore on an empty cluster, otherwise restore will fail because Cluster API is not installed.
Am I wrong?

Reply:
Just to clarify, in restore CRD should exist in the target cluster, but they are there...

Reply:
Ah I see - that makes sense. So, if I'm understanding, the `getDiscoveryTypes` method should reach out to a target management cluster (be it in a move or a restore), and ensure that the Cluster API CRDs are available, present? 

I think your assumption is correct that in a restore, the CRDs _should_ already be there before a restore is attempted

Reply:
I think we should make getDiscoveryTypes flexible enough to reach out for the source or the target cluster, and use the latter only for restore. Another option a little bit less explicit is to enforce source=target for restore

Reply:
I did a little bit more of digging in this PR and in Restore and I L154 we are doing
```
objectGraph := newObjectGraph(o.fromProxy, o.fromProviderInventory)
```
also in L200 we are doing
```
if err := o.restore(objectGraph, o.fromProxy); err != nil {
```
That's makes me think that o.fromProxy in fact is the target proxy for restore, which is kind of confusing.
What about passing `toCluster Client`, explicitly to restore, like we are doing for move?